src/worker/jobs/expiry_notification.rs

```diff
@@ -0,0 +1,229 @@
+use crate::models::ApiToken;
+use crate::schema::api_tokens;
+use crate::{email::Email, models::User, worker::Environment, Emails};
+use anyhow::anyhow;
+use chrono::SecondsFormat;
+use crates_io_worker::BackgroundJob;
+use diesel::dsl::IntervalDsl;
+use diesel::{
+    dsl::now, BoolExpressionMethods, Connection, ExpressionMethods, NullableExpressionMethods,
+    PgConnection, QueryDsl, QueryResult, RunQueryDsl, SelectableHelper,
+};
+use std::sync::Arc;
+
+/// The threshold in days for the expiry notification.
+const EXPIRY_THRESHOLD: chrono::TimeDelta = chrono::TimeDelta::days(3);
+
+/// The maximum number of tokens to check.
+/// To avoid sending too many emails and submitting to many transactions, we limit the number of
+/// tokens to check.
+const MAX_ROWS: i64 = 10000;
+/// A job responsible for monitoring the status of a token.
+/// It checks if the token is about to reach its expiry date.
+/// If the token is about to expire, the job triggers a notification.
+#[derive(Default, Serialize, Deserialize, Debug)]
+pub struct SendTokenExpiryNotifications;
+
+impl BackgroundJob for SendTokenExpiryNotifications {
+    const JOB_NAME: &'static str = "expiry_notification";
+
+    type Context = Arc<Environment>;
+
+    #[instrument(skip(env), err)]
+    async fn run(&self, env: Self::Context) -> anyhow::Result<()> {
+        let conn = env.deadpool.get().await?;
+        conn.interact(move |conn| {
+            // Check if the token is about to expire
+            // If the token is about to expire, trigger a notification.
+            check(&env.emails, conn)
+        })
+        .await
+        .map_err(|err| anyhow!(err.to_string()))?
+    }
+}
+// Check if the token is about to expire and send a notification if it is.
+fn check(emails: &Emails, conn: &mut PgConnection) -> anyhow::Result<()> {
+    info!("Checking if tokens are about to expire");
+    let expired_tokens = find_tokens_expiring_within_days(conn, EXPIRY_THRESHOLD)?;
+    if expired_tokens.len() == MAX_ROWS as usize {
+        warn!("The number of tokens to check exceeds the limit");
+    }
+    for token in &expired_tokens {
+        conn.transaction(|conn| {
```

Comment:
we can probably get rid of the transaction block now since there is only a single write in the transaction and no following reads. might make sense to extract a `fn handle_expiring_token(emails, conn, token) { ... }` for that purpose so that we can use `?` inside the fn and only have to write the error handling code once.